[20849] Reorganize and speed up test suite

basiszwo opened the pull request:
https://community.openproject.org/work_packages/20849

I did some reorganizing and broke up the test matrix. We are down to 20mins 
runtime. In case we can get more parallel jobs to run (currently 5) we can
decrease even more. 
#### What has been done?
- do not ignore schema.rb and make use of it
- generate coverage report only on demand
- remove object daddy and fully switch to factory girl
- largely split test matrix
- only test against postgresql (myssql should not be checked on each run)
- fix rspec require statements
- extract rspec configs to separate files
- update rspec config
- run most of the specs with capybara-webkit except selenium dependent ones
- reorganize rspec files
  - combine api specs
  - divide rspec feature specs in modules (1st step towards modules and required for build matrix)
- reorganize cucumber features: divide into modules (mainly for build matrix)
- add license to some files
- add parallel_test gem
- require features in cucumber config to be able to run `cucumber features/foo/bar.feature`
- do not create screenshots for failing specs in CI
#### Next steps
- Test parallel_test on travis and see if this decreases test time even more.
- run webpack only once
- compile asset before
- identify slowest tests and rewrite them

basiszwo committed:
Add capybara-webkit gem

basiszwo committed:
Use capybara webkit as default javascript driver

This will break specs so it is not a green commit.

basiszwo committed:
Fix specs that rely on selenium features

Add a custom metadata tag to run specs with selenium instead of
capybara webkit. As some next steps we need to bring in the required
functionality to capybara webkit to get rid of (slow) selenium.

basiszwo committed:
Remove rspec exclusion instruction as legacy specs have been removed

basiszwo committed:
Generate code coverage report selectively

basiszwo committed:
Require factory girl manually to fix db:schema:load error

Within the factories a lot of factory girl uncommon stuff goes on.
Instantiate classes, build other factories, ...
This results in ActiveRecord errors when the database gets set up via
rake db:schema:load

basiszwo committed:
Use factory girl instead of object daddy

basiszwo committed:
Only test postgres specs and cukes

basiszwo committed:
Include db/schema.rb

basiszwo committed:
Only run specs (no cukes)

basiszwo committed:
Remove object daddy and its exemplars

basiszwo committed:
Divide test matrix in feature / no_feature specs

basiszwo committed:
Split build matrix for travis

basiszwo committed:
Add hint to webpack stuff only before running feature specs

basiszwo committed:
Streamline rspec require statements

basiszwo committed:
Remove testing code for legacy specs and test unit

basiszwo committed:
Further split travis build matrix

basiszwo committed:
Add rake task to run all specs except slower ones

misc task runs everything except controller, requests, features and
model specs

basiszwo committed:
Extract rspec config for capybara-screenshot

Do not create screenshots on ci servers

basiszwo committed:
Extract rspec config for factory girl

basiszwo committed:
Extract rspec config for database cleaner

basiszwo committed:
Extract rspec config for capybara

basiszwo committed:
Extract rspec conig for openproject cleanup tasks

basiszwo committed:
Extract loading of disabled specs from openproject plugins

basiszwo committed:
Extract rspec config for angularjs dsl

basiszwo committed:
Extract rspec config for json spec helpers

basiszwo committed:
View specs need webacked assets

basiszwo committed:
Rename capybara selenium fallback file

basiszwo committed:
Do not test failing and unfixed rspec features (tmp)

basiszwo committed:
Extract and combine all api related specs

basiszwo committed:
Split travis build matrix for api specs

basiszwo committed:
Move api v3 feature spec

basiszwo committed:
Keep log folder

basiszwo committed:
Upgrade capybara and update capybara config

basiszwo committed:
Identify selenium dependent feature specs

basiszwo committed:
Combine work package page objects and contexts

basiszwo committed:
Add missing license in work package feature specs

basiszwo committed:
Run rspec features on travis

basiszwo committed:
Add parallel tests gem

basiszwo committed:
Use db:schema:load in CI

basiszwo committed:
Increase capybara timeout to 5 seconds

basiszwo committed:
Brake down api specs to different versions in travis build matrix

basiszwo committed:
Split api v3 tasks for faster build matrix on travis

basiszwo committed:
Move tableless spec helper to support folder

basiszwo committed:
Require features to run single cuke with cucumber command

basiszwo committed:
Define cucumber:matrix task for travis build matrix

basiszwo committed:
Remove object daddy requirement for cukes

basiszwo committed:
Divide cucumber features into different namespace folders

basiszwo committed:
Run cucumber namespaces to check which ones are capybara incompatible

basiszwo committed:
Do not create screenshot on cuke fail in CI

basiszwo committed:
Tag selenium dependent cucumber features

basiszwo committed:
Reorder travis build matrix

basiszwo committed:
Do not run @wip tasks

basiszwo committed:
Accept dialog in work package attachment cucumber feature

basiszwo committed:
Tag some cucumber features as wip

basiszwo committed:
Increase capybara wait time for cucumber features to 5 secs

basiszwo committed:
Run complete travis test matrix

basiszwo committed:
Add new feature namespace 'admin'

basiszwo committed:
Run admin cucumber features in build matrix

basiszwo committed:
Add note to api v1 specs

basiszwo committed:
Use selenium to run top menu item specs

basiszwo committed:
Extract rspec specific tasks to separate rake file

basiszwo committed:
Move api v3 specs v3 folder

basiszwo committed:
Create subfolders for specs

This is mainly done because of 2 reasons:
1. we want to split OpenProject in modules anyway
2. we can speed up the test matrix by separating the specs

I think this separation needs more refinement which will be applied
in future commits.

basiszwo committed:
Properly define rspec rake tasks for api and feature subfolders

We want to have proper namespaced rake tasks which we can use
for the travis test matrix and to run some specs in isolation.

basiszwo committed:
Upgrade rspec config

This splits the helpers into spec_ and rails_helper. We need to require
rails_helper in all specs.

basiszwo committed:
Use selenium for admin user feature

basiszwo committed:
Reorganize spec/features/work_packages and split travis build matrix

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://basiszwo-20849-test-suite.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://basiszwo-20849-test-suite.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented:
> Remove legacy specs as they are not being run anyway

This is not true. See Travis.

Azure7111 commented on .travis.yml:
```diff
@@ -79,9 +100,7 @@ script:
-  webhooks:
-    urls:
-      - "https://webhooks.gitter.im/e/435ab5d5c6944305da2f"
+  slack: openprojectservices:VF0xFQjYrBWDMFveAtUtxJDL
```
There is a slack now for OpenProejct?

ulferts commented:
I am pretty sure that e.g. https://github.com/opf/openproject/blob/dev/spec/legacy/functional/user_mailer_spec.rb covers code we haven't covered otherwise. I'd therefore like to process a tad more selectively when it comes to removing specs. 

We could for example check whether we at least have a corresponding file in our non legacy test suite before removing a file from the legacy suite.

basiszwo committed:
Properly require factory girl

basiszwo committed:
Decouple legacy specs from ObjectDaddy

basiszwo committed:
Run legacy specs in build matrix

basiszwo committed:
Remove clearing of rake spec and rake spec:legacy

basiszwo committed:
Fixme note that reload is being overwritten for user

basiszwo committed:
Reactivate gitter hook

ulferts commented:
This is a pretty big PR bringing Github to it's limit. I am therefore refraining from commenting on individual changes and rather write down what I noticed in one comment. It might make sense to break parts of this PR out into smaller ones and squash other changes to make it reviewable.

# I like
- Splitting up the tests into smaller chunks should make it easier to execute them separately and also it should give some more structure. Maybe we'll one day be able to break chunks of the application out of here completely. That would be the most efficient speedup we can get.
- Structuring the tests by functional groups (e.g. work_package, wiki)  
- Cleaning up the codebase (structure and removal of old dependencies).

# I wish
- That the breaking up into functional groups would also be applied to the non feature specs. We now still have a huge glob of model/controller/view specs. They'll need to be broken up into smaller pieces still. And IMO it should also be by feature. Is a big pile of work though, so I already appreciate what you have done so far.
- We'd also enforce breaking up by the spec/test_helper that is available within each block. We could have a test helper for tests that only supports unit tests in their strict definition (No DB/FS access), one for unit tests requiring DB access, one for integrational/functional (in the sense of testing at least two classes working together) and one for E2E tests. Those would then only load the parts of the application that is required for running the particular type of test leading to fast ramp up time. Additionally, we would force developers into a schema within each test type.  
- That the `@wip` tags added where removed again. 

# I fear
- The overhead in setup time when breaking up the tests in smaller chunks like this especially on the internal CI. 
- That it will take additional effort to make it work on the internal CI with the configurations (OP core + plugins) we maintain. We'll have to handle the different schema.rb per configuration.
- That running the tests only on postgresql will lead to failures being discovered by our internal CI after they have been merged in.
- That the removal of snapshots on the CI will make it harder to find flickering specs. That feature was introduced particularly to understand such failures. If they are to be disabled, they should be disabled when not running on a CI. I don't care for them locally as I can debug. 

# Additional notes
- There are references to parallel_test in there although it is not used yet. Do you plan on employing it within this PR? 
- Before merging, this needs to be executed on the internal CI for every configuration we have at least once.

NobodysNightmare commented:
@basiszwo wrote:

> only test against postgresql (myssql should not be checked on each run)

@ulferts wrote:

> I fear [...] That running the tests only on postgresql will lead to failures being discovered by our internal CI after they have been merged in.

I agree with the fear of ulferts. Code that is not fully DB agnostic is very well possible. Just today there was an example of code only working on PSQL and not for MySQL, I myself have written code taht failed for PSQL, but worked on MySQL. I see that those double runs are a bit of a pain, but at the same time I feel they are neccessary.

When not asking **if** we want to disable one of two adapters for travis runs, I would be asking **which** adapter to disable. Is there a rationale to choose PSQL over MySQL? Or is that a decision based on a gut feeling?

NobodysNightmare commented on Gemfile.lock:
```diff
@@ -590,6 +592,3 @@ DEPENDENCIES
   will_paginate (~> 3.0)
-
-BUNDLED WITH
-   1.10.6
```
@basiszwo Please update your bundler version. That this line is removed is an annoyance that the bundler team describes here:

http://bundler.io/blog/2015/06/24/version-1-10-released.html#bundled-with

**tl;dr:** New bundlers will add this line, old bundlers will remove it: **Git edit wars!** Update to a recent bundler: Worst case of one change per bundler version.

edit: for reference: the offending commit is https://github.com/basiszwo/openproject/commit/30a559da75e81d50d5af59c9bd51cf5ac8c8eee4

NobodysNightmare commented on lib/tasks/ci.rake:
```diff
@@ -81,10 +81,11 @@ namespace :ci do
+      # Rake::Task['db:load_config'].execute
+      #
+      # Rake::Task['db:migrate'].invoke
+      # Rake::Task['db:schema:dump'].invoke
```
Can we safely delete those commented lines?

NobodysNightmare commented:
Something I noticed for the first time:

> [DEPRECATION] Capybara::Webkit::Driver#browser is deprecated.

Is this new to your PR? I'd say yes. Found here: https://travis-ci.org/opf/openproject/jobs/75937099

(There are other deprecation warnings in our tests... not saying those are the first deprecation warnings at all)

basiszwo committed:
Fix previously wip'ed cucumber features

basiszwo committed:
Show me the page now creates a screenshot

basiszwo committed:
Resize window to 1680x1000 when using selenium driver

basiszwo committed:
Remove commented code in ci.rake

basiszwo committed:
Fix describe statements in legacy specs

basiszwo committed:
Add license to spec/support files

basiszwo committed:
Reorganize legacy spec helper

basiszwo committed:
Do not run legacy specs in spec:misc task

basiszwo committed:
Run qt4 dependent features with selenium

Travis does not yet support qt5 which is required to pass these
features with capybara webkit.

basiszwo committed:
Add note about rspec legacy formatters

basiszwo committed:
Cucumber features for core are already loaded in cucumber.yml

In cucumber.yml we already load the core cucumber features so we don't
want to load them again as this causes Cucumber::Ambiguous errors.

basiszwo committed:
Properly align block

basiszwo commented:
@NobodysNightmare as for the deprecation warning see https://github.com/thoughtbot/capybara-webkit/pull/778 I hope to get rid of the warnings soon.

basiszwo committed:
Define rspec:legacy task manually to not run spec:prepare

basiszwo committed:
Temporarily disable failing legacy specs

These specs fail when run in total but are green when run solo.

basiszwo commented:
closing this pull request for further testing that should not pollute the official build pipeline.
